File: cpp/85.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  int maximalRectangle(vector<vector<char> > &matrix) {
    int m = matrix.size();
    if(!m) return 0;
    int n = matrix[0].size();
    int dp[n][n];
    for(int i=0; i<n; i++)
      for(int j=0; j<n; j++)
        dp[i][j] = 0;
    int ans = 0;
    for(int k=0; k<m; k++){
      for(int i=0; i<n; i++)
        for(int j=i; j<n; j++){
          if(i == j)
            dp[i][j] = matrix[k][j] == '1' ? dp[i][j]+1 : 0;
          else
            dp[i][j] = dp[i][j-1] && matrix[k][j] == '1' ? dp[i][j]+1 : 0;
          ans = max(ans, dp[i][j] * (j-i+1));
        }
    }
    return ans;
  }
};

int main(){
  auto sol = Solution();
  vector<vector<char> > m = {{'0', '0'}};
  cout << sol.maximalRectangle(m);
}

File: python/85.py
class Solution(object):
    # from 84
    def largestRectangleArea(self, height):
        n = len(height)
        stack = []
        lefts = []
        rights = [n] * n
        for i, h in enumerate(height):
            while stack and height[stack[-1]] > h:  # >=
                rights[stack.pop()] = i             # tricky!
            lefts.append(-1 if not stack else stack[-1])
            stack.append(i)
        ans = 0
        for h, l, r in zip(height, lefts, rights):
            ans = max(ans, (r-l-1) * h)
        return ans

    def maximalRectangle(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        n = len(matrix)
        if not n:
            return 0
        m = len(matrix[0])
        if not m:
            return 0
        height = [0] * m
        ans = 0
        for i in xrange(n):
            for j in xrange(m):
                height[j] = 0 if matrix[i][j] == '0' else 1 + height[j]
            ans = max(ans, self.largestRectangleArea(height))
        return ans
